lib/aws/client.ex

```diff
@@ -194,12 +194,83 @@ defmodule AWS.Client do
     %{client | http_client: http_client}
   end
 
+  @doc """
+  `opts` may contain the keyword `:enable_retries?` that enables request retries on known
+  errors such as 500s.
+  The keyword `:retry_opts` can further supply the arguments: `:max_retries`, `:base_sleep_time` &
+  `:cap_sleep_time` that control the retry mechanism. This uses exponential backoff with jitter to
+  sleep in between retry attempts. Default values are:
+  `retry_opts: [max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000]`
+
+  See "FullJitter" at: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+  """
   def request(client, method, url, body, headers, opts \\ []) do
+    {enable_retries?, opts} = Keyword.pop(opts, :enable_retries?, false)
+    {retry_num, opts} = Keyword.pop(opts, :retry_num, 0)
+    {retry_opts, opts} = Keyword.pop(opts, :retry_opts, nil)
+
     {mod, options} = Map.fetch!(client, :http_client)
     options = Keyword.merge(options, opts)
-    apply(mod, :request, [method, url, body, headers, options])
+    resp = apply(mod, :request, [method, url, body, headers, options])
+
+    retriable?(resp)
+    |> case do
+      :ok ->
+        resp
+
+      :retry ->
+        if enable_retries? and should_retry?(retry_num, retry_opts) do
+          updated_opts =
+            Keyword.merge(opts,
+              retry_num: retry_num + 1,
+              enable_retries?: enable_retries?,
+              retry_opts: retry_opts
+            )
+
+          request(client, method, url, body, headers, updated_opts)
+        else
+          resp
+        end
+
+      :error ->
+        resp
+    end
   end
 
+  def should_retry?(_retry_num, nil), do: false
```

Comment:
Hmm won't this now still result in `false` if `:enable_retries?` is set to `true` but no specific `retry_opts` are provided? (Hence I want to rely on the defaults). 

I think it can be simplified if you were to move the `retry_opts` defaults further up:
```elixir
{retry_opts0, opts} = Keyword.pop(opts, :retry_opts, [])
retry_opts = Keyword.merge([max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000], retry_opts0)
```

In that case this function clause simply goes out the window and the `Keyword.get/3` calls below here could become `{:ok, some_variable_name} = Keyword.fetch(retry_opts, :key)` to avoid having the defaults in two places.

Reply:
You are absolutely right again! I used your suggestion, and cleaned up the `Keyword.get/3` further below. Added a test for that specific case too.